libslide/src/evaluator_rules/pattern_matching.rs

```diff
@@ -138,42 +165,93 @@ mod tests {
 
     mod match_rule {
         use super::*;
+        use crate::{parse, scan, ScannerOptions};
 
-        fn var(s: &str) -> Expr {
-            Expr::Var(Var {
-                name: s.to_string(),
-            })
-        }
-        fn const_e(n: f64) -> Expr {
-            Expr::Const(n)
-        }
+        macro_rules! match_rule_tests {
+            ($($name:ident: $rule:expr => $target:expr => $expected_repls:expr)*) => {
+            $(
+                #[test]
+                fn $name() {
+                    let parse = |prog: &str, is_rule: bool| -> Expr {
+                        let mut scanner_options = ScannerOptions::default();
+                        if is_rule {
+                            scanner_options = scanner_options.set_is_var_char(|c| {
+                                c.is_alphabetic() || c == '$' || c == '#' || c == '_'
+                            });
+                        }
+                        match parse(scan(prog, scanner_options)) {
+                            Stmt::Expr(expr) => expr,
+                            _ => unreachable!(),
+                        }
+                    };
 
-        // TODO: more tests?
-        #[test]
-        fn match_rule_has_match() {
-            let replacements = match_rule(
-                Expr::BinaryExpr(BinaryExpr {
-                    op: BinaryOperator::Plus,
-                    lhs: var("$a").into(),
-                    rhs: const_e(0.).into(),
-                }),
-                Expr::BinaryExpr(BinaryExpr {
-                    op: BinaryOperator::Plus,
-                    lhs: var("x").into(),
-                    rhs: const_e(0.).into(),
-                }),
-            )
-            .unwrap();
-
-            assert_eq!(replacements.map.len(), 1);
-            assert_eq!(replacements.map.get(&var("$a")).unwrap().to_string(), "x");
+                    let parsed_rule = parse($rule, true);
+                    let parsed_target = parse($target, false);
+
+                    let repls = match_rule(parsed_rule, parsed_target);
+                    let (repls, expected_repls): (Replacements, Vec<&str>) =
+                        match (repls, $expected_repls) {
+                            (None, expected_matches) => {
+                                assert!(expected_matches.is_none());
+                                return;
+                            }
+                            (Some(repl), expected_matches) => {
+                                assert!(expected_matches.is_some());
+                                (repl, expected_matches.unwrap())
```

Comment:
won't unwrap through here if there is nothing so we don't need the assert?

Reply:
This is just to make the error messages a little bit nicer. If it panics, we just get a stack trace and it's not super clear where the error happened. With the assertion we're told where the error took place.

Reply:
gotcha. smart.